aspnetcore/web-api/action-return-types.md

```diff
@@ -10,6 +10,188 @@ uid: web-api/action-return-types
+
+This table shows the common result helpers available: [Built-in results](/aspnet/core/fundamentals/minimal-apis#built-in-results-1)
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/IResultProductsController.cs" id="snippet_GetByIdIResult" highlight="7":::
+
+In the preceding action:
+
+* A 404 status code is returned when the product doesn't exist in the database.
+* A 200 status code is returned with the corresponding `Product` object when the product does exist, generated by the [Results.Ok\<T>()](/dotnet/api/microsoft.aspnetcore.http.results.ok).
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/IResultProductsController.cs" id="snippet_CreateAsyncIResult" highlight="9,16":::
+
+In the preceding action:
+
+* A 400 status code is returned by the ASP.NET Core runtime when:
+  * The [`[ApiController]`](xref:Microsoft.AspNetCore.Mvc.ApiControllerAttribute) attribute has been applied and model validation fails.
+  * The product description contains "XYZ Widget".
+* A 201 status code is generated by the [`Results.Create`](/dotnet/api/microsoft.aspnetcore.http.results.created) method when a product is created. In this code path, the `Product` object is provided in the response body. A `Location` response header containing the newly created product's URL is provided.
+
+### Result\<T> type
+
+ASP.NET Core also includes the static [TypedResults](<xref:Microsoft.AspNetCore.Http.TypedResults>) class that returns the concrete `IResult` implementation, that allows using it as return type. The usage of the concrete `IResult` implementation offers the following benefit over the [IResult type](#iresult-type):
+
+* All the [`[ProducesResponseType]`](xref:Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute) attribute's can be excluded, since the `HttpResult` implementation will contribute automatically to the endpoint metadata.
+
+When multiple `IResult` return types are needed, while still possible use `IResult` interface as return type is better use the [`Result<TResult1, TResultN>`](/dotnet/api/microsoft.aspnetcore.http.httpresults.results-2) generic union types that will keep the automatic endpoint metadata benefit.
+
+The `Results<TResult1, TResultN>` union types implement implicit cast operators so that the compiler can automatically convert the types specified in the generic arguments to an instance of the union type. This has the added benefit of providing compile-time checking that a route handler actually only returns the results that it declares it does. Attempting to return a type that isn’t declared as one of the generic arguments to `Results<>` will result in a compilation error.
```

Comment:
```suggestion
The `Results<TResult1, TResultN>` union types implement implicit cast operators so that the compiler can automatically convert the types specified in the generic arguments to an instance of the union type. This has the added benefit of providing compile-time checking that a route handler actually only returns the results that it declares it does. Attempting to return a type that isn’t declared as one of the generic arguments to `Results<>` results in a compilation error.
```

Consider the following code: